File: Packages/TeamTasksKit/Sources/TeamTasksContract/Verify.swift
import Foundation
import TeamTasksCore

/// A failed contract expectation, with its source location.
public struct ContractFailure: Error, CustomStringConvertible, Sendable {
    public let message: String
    public let location: String

    public init(_ message: String, file: StaticString = #fileID, line: UInt = #line) {
        self.message = message
        location = "\(file):\(line)"
    }

    public var description: String { "\(message) (\(location))" }
}

extension ContractFailure: LocalizedError {
    public var errorDescription: String? { description }
}

/// Tiny assertion helpers that throw descriptive `ContractFailure`s (no dependency on a test framework).
public enum Verify {
    public static func that(
        _ condition: Bool,
        _ message: @autoclosure () -> String,
        file: StaticString = #fileID,
        line: UInt = #line
    ) throws {
        guard condition else { throw ContractFailure(message(), file: file, line: line) }
    }

    public static func equal<Value: Equatable>(
        _ actual: Value,
        _ expected: Value,
        _ message: @autoclosure () -> String,
        file: StaticString = #fileID,
        line: UInt = #line
    ) throws {
        guard actual == expected else {
            throw ContractFailure("\(message()): expected \(expected), got \(actual)", file: file, line: line)
        }
    }

    public static func unwrap<Value>(
        _ value: Value?,
        _ message: @autoclosure () -> String,
        file: StaticString = #fileID,
        line: UInt = #line
    ) throws -> Value {
        guard let value else { throw ContractFailure("\(message()): unexpected nil", file: file, line: line) }
        return value
    }

    /// Expects `body` to throw exactly `expected`.
    public static func fails<Value>(
        with expected: AppError,
        _ message: @autoclosure () -> String,
        file: StaticString = #fileID,
        line: UInt = #line,
        _ body: () async throws -> Value
    ) async throws {
        try await fails(withAnyOf: [expected], message(), file: file, line: line, body)
    }

    /// Expects `body` to throw one of `expected`.
    public static func fails<Value>(
        withAnyOf expected: Set<AppError>,
        _ message: @autoclosure () -> String,
        file: StaticString = #fileID,
        line: UInt = #line,
        _ body: () async throws -> Value
    ) async throws {
        let wanted = expected.map { "\($0)" }.sorted().joined(separator: " | ")
        do {
            let value = try await body()
            throw ContractFailure("\(message()): expected error \(wanted), got success \(value)", file: file, line: line)
        } catch let error as ContractFailure {
            throw error
        } catch let error as AppError {
            guard expected.contains(error) else {
                throw ContractFailure("\(message()): expected error \(wanted), got \(error)", file: file, line: line)
            }
        } catch {
            throw ContractFailure("\(message()): expected error \(wanted), got non-AppError \(error)", file: file, line: line)
        }
    }

    /// Non-members see nothing: an empty result (RLS), never an error (lead decision).
    public static func hidden<Result: Collection>(
        _ message: @autoclosure () -> String,
        file: StaticString = #fileID,
        line: UInt = #line,
        _ body: () async throws -> Result
    ) async throws {
        let result: Result
        do {
            result = try await body()
        } catch let error as ContractFailure {
            throw error
        } catch {
            throw ContractFailure("\(message()): expected an empty result, got error \(error)", file: file, line: line)
        }
        guard result.isEmpty else {
            throw ContractFailure("\(message()): expected nothing visible, got \(result)", file: file, line: line)
        }
    }

    /// Runs a setup/action step and labels any unexpected error with `label`.
    @discardableResult
    public static func step<Value>(
        _ label: @autoclosure () -> String,
        file: StaticString = #fileID,
        line: UInt = #line,
        _ body: () async throws -> Value
    ) async throws -> Value {
        do {
            return try await body()
        } catch let error as ContractFailure {
            throw error
        } catch {
            throw ContractFailure("\(label()) failed: \(error)", file: file, line: line)
        }
    }
}

File: android/contract/src/main/kotlin/io/github/notkanaa/equipe/contract/Verify.kt
package io.github.notkanaa.equipe.contract

import io.github.notkanaa.equipe.core.AppError
import kotlin.coroutines.cancellation.CancellationException

/**
 * A failed contract expectation. [location] is the scenario source line that raised it (`File.kt:line`, best effort
 * from the stack trace: the first frame outside the helpers of this package), like Swift's `#fileID:#line`.
 */
class ContractFailure(override val message: String) : Exception(message) {
    val location: String by lazy { callSite(stackTrace) }

    /** `message (location)`. */
    val description: String get() = "$message ($location)"

    override fun toString(): String = description

    private companion object {
        private val helperClasses = listOf(
            ContractFailure::class.java.name,
            Verify::class.java.name,
            StreamProbe::class.java.name,
            RealtimeProbe::class.java.name,
        )

        fun callSite(frames: Array<StackTraceElement>): String {
            val frame = frames.firstOrNull { frame ->
                frame.fileName != null && frame.lineNumber > 0 &&
                    frame.className.startsWith("io.github.notkanaa.equipe.") &&
                    helperClasses.none { frame.className == it || frame.className.startsWith("$it$") }
            }
            return if (frame == null) "?" else "${frame.fileName}:${frame.lineNumber}"
        }
    }
}

/** Tiny assertion helpers that throw descriptive [ContractFailure]s (no dependency on a test framework). */
object Verify {
    fun that(condition: Boolean, message: String) {
        if (!condition) throw ContractFailure(message)
    }

    fun <T> equal(actual: T, expected: T, message: String) {
        if (actual != expected) throw ContractFailure("$message: expected $expected, got $actual")
    }

    fun <T : Any> unwrap(value: T?, message: String): T = value ?: throw ContractFailure("$message: unexpected null")

    /** Expects [body] to throw exactly [expected]. */
    suspend fun fails(expected: AppError, message: String, body: suspend () -> Any?) {
        failsWithAnyOf(setOf(expected), message, body)
    }

    /** Expects [body] to throw one of [expected]. */
    suspend fun failsWithAnyOf(expected: Set<AppError>, message: String, body: suspend () -> Any?) {
        val wanted = expected.map { "$it" }.sorted().joinToString(" | ")
        val value = try {
            body()
        } catch (error: ContractFailure) {
            throw error
        } catch (error: CancellationException) {
            throw error
        } catch (error: AppError) {
            if (error !in expected) throw ContractFailure("$message: expected error $wanted, got $error")
            return
        } catch (error: Throwable) {
            throw ContractFailure("$message: expected error $wanted, got non-AppError $error")
        }
        throw ContractFailure("$message: expected error $wanted, got success $value")
    }

    /** Non-members see nothing: an empty result (RLS), never an error (lead decision). */
    suspend fun hidden(message: String, body: suspend () -> Collection<*>) {
        val result = try {
            body()
        } catch (error: ContractFailure) {
            throw error
        } catch (error: CancellationException) {
            throw error
        } catch (error: Throwable) {
            throw ContractFailure("$message: expected an empty result, got error $error")
        }
        if (result.isNotEmpty()) throw ContractFailure("$message: expected nothing visible, got $result")
    }

    /** Runs a setup/action step and labels any unexpected error with [label]. */
    suspend fun <T> step(label: String, body: suspend () -> T): T = try {
        body()
    } catch (error: ContractFailure) {
        throw error
    } catch (error: CancellationException) {
        throw error
    } catch (error: Throwable) {
        throw ContractFailure("$label failed: $error")
    }
}
